### PR TITLE
DAO Interface Major Changes

### PR DESCRIPTION
**Pull request by kennyqiyenkan**
# DAO Interface Change
This PR consists of changes to the DAO Interface as discussed in this [discussion](https://groups.google.com/forum/#!msg/foam-framework-discuss/LTT2dhADcJc/EbDgcMDYCAAJ).

### CODE REVIEW

File: java_src/foam/dao/DatastoreDAO.java [hunk 1]
```diff
@@ -132,7 +133,7 @@ protected FObject deserialize(PropertyContainer e) {
         return obj;
     }
 
-    public Sink select(Sink sink, Integer skip, Integer limit, Comparator order, Predicate predicate) {
+    public Sink select_(X x, Sink sink, Integer skip, Integer limit, Comparator order, Predicate predicate) {
```
Comment (kgrgreer) [inline java_src/foam/dao/DatastoreDAO.java]: Integer -> int

Integers create objects on the heap and so are more expensive to build and need to be GC'ed.

File: java_src/foam/dao/DatastoreDAO.java [hunk 2]
```diff
@@ -160,10 +161,10 @@ public Sink select(Sink sink, Integer skip, Integer limit, Comparator order, Pre
         return sink;
     }
 
-    public void removeAll(Integer skip, Integer limit, Comparator order, Predicate predicate) {
+    public void removeAll_(X x, Integer skip, Integer limit, Comparator order, Predicate predicate) {
```
Comment (kgrgreer) [inline java_src/foam/dao/DatastoreDAO.java]: Integer->int, or better long

Comment (kennyqiyenkan) [reply java_src/foam/dao/DatastoreDAO.java]: shall I change all Integers in the DAO to become longs then?

File: java_src/foam/dao/JournaledDAO.java [hunk 3]
```diff
@@ -29,31 +30,31 @@ public JournaledDAO(DAO delegate, String filename) throws IOException {
    * @returns FObject
    */
   @Override
-  public FObject put(FObject obj) {
+  public FObject put_(X x, FObject obj) {
     journal.put(obj, null);
```
Comment (kgrgreer) [inline java_src/foam/dao/JournaledDAO.java]: Call put_

File: java_src/foam/dao/JournaledDAO.java [hunk 4]
```diff
@@ -29,31 +30,31 @@ public JournaledDAO(DAO delegate, String filename) throws IOException {
...
-    return getDelegate().put(obj);
+    return getDelegate().put_(x, obj);
   }
 
   @Override
-  public FObject remove(FObject obj) {
+  public FObject remove_(X x, FObject obj) {
     Object id = ((AbstractDAO) getDelegate()).getPrimaryKey().get(obj);
     journal.remove(id, null);
```
Comment (kgrgreer) [inline java_src/foam/dao/JournaledDAO.java]: call remove_

File: java_src/foam/dao/JournaledDAO.java [hunk 5]
```diff
@@ -29,31 +30,31 @@ public JournaledDAO(DAO delegate, String filename) throws IOException {
...
   @Override
-  public Sink select(Sink sink, Integer skip, Integer limit, Comparator order, Predicate predicate) {
+  public Sink select_(X x, Sink sink, Integer skip, Integer limit, Comparator order, Predicate predicate) {
     return null;
   }
 
   @Override
-  public void removeAll(Integer skip, Integer limit, Comparator order, Predicate predicate) {
+  public void removeAll_(X x, Integer skip, Integer limit, Comparator order, Predicate predicate) {
     journal.removeAll();
```
Comment (kgrgreer) [inline java_src/foam/dao/JournaledDAO.java]: Call removeAll_().

File: src/foam/dao/LoggingDAO.js [hunk 6]
```diff
@@ -62,21 +62,21 @@ foam.CLASS({
...
       }
-      return this.SUPER(sink, skip, limit, order, predicate);
+      return this.SUPER(x, sink, skip, limit, order, predicate);
     },
 
-    function removeAll(sink, skip, limit, order, predicate) {
-      this.logger('removeAll', skip, limit, order, predicate);
-      return this.SUPER(sink, skip, limit, order, predicate);
+    function removeAll_(x, sink, skip, limit, order, predicate) {
+      this.logger('removeAll_', skip, limit, order, predicate);
```
Comment (kgrgreer) [inline src/foam/dao/LoggingDAO.js]: Log as 'removeAll' not 'removeAll_'.

File: src/foam/dao/MDAO.js [hunk 7]
```diff
@@ -156,19 +156,19 @@ foam.CLASS({
...
       }
 
-      return Promise.resolve(this.find_(
+      return Promise.resolve(this.find__(
           this.of.isInstance(objOrKey) ? objOrKey.id : objOrKey));
     },
 
     /** internal, synchronous version of find, does not throw */
-    function find_(key) {
+    function find__(key) {
```
Comment (kgrgreer) [inline src/foam/dao/MDAO.js]: find, find_, find__ is too confusing, rename to findSync_

File: java_src/foam/dao/JournaledDAO.java (hunk 5 above)
Comment (kennyqiyenkan) [reply java_src/foam/dao/JournaledDAO.java]: calling removeAll on journal would mean I would be refactoring a non DAO object. If we want to make this change, I have to change `FileJournal.java` as well.

Edit: Actually, it goes all the way to the Sink level. So this change, as well as remove_ and put_ as you've commented above, would require a refactor of Sink.

Comment (kgrgreer) [reply java_src/foam/dao/JournaledDAO.java]: My mistake.

File: src/foam/dao/LRUDAOManager.js [hunk 8]
```diff
@@ -102,7 +102,7 @@ foam.CLASS({
   listeners: [
     /** Adds the put() item to the tracking dao, runs cleanup() to check
       the dao size. */
-    function onPut(s, on, put, obj) {
+    function onPut(s, on, put_, obj) {
```
Comment (kgrgreer) [inline src/foam/dao/LRUDAOManager.js:105]: Rename put_ to put.

**Pull request merged**